offsetlog.go

```diff
@@ -0,0 +1,197 @@
+package kafka
+
+import (
+	"context"
+	"errors"
+	"math/rand"
+	"time"
+)
+
+// offsetLog is an implementation of the offsetStore interface which uses a
+// Kafka partition as backing store for the offsets.
+//
+// Ideally the partition belongs to a compacted topic, but this is not required
+// and the log will work just fine with a regular topic (replays may take a bit
+// longer if there are large numbers of offsets recorded in the log).
+type offsetLog struct {
+	// Addresses of the Kafka brokers tha the log uses to lookup the leader of
+	// the topic partition that it uses.
+	brokers []string
+
+	// Name of the Kafka topic that the offset log uses to keep track of the
+	// offsets.
+	topic string
+
+	// Partition tha the offset log uses in the Kafka topic.
+	partition int
+
+	// Size limit of the fetch requests to the offset log, default to 8 MB.
+	maxBytes int
+
+	// Time limit applied to operations of the offset log.
+	readTimeout  time.Duration
+	writeTimeout time.Duration
+
+	// Dialer used by the offset log to establish connections to the Kafka
+	// brokers. If nil, DefaultDialer is used.
+	dialer *Dialer
+}
+
+const (
+	defaultOffsetLogReadTimeout  = 3 * time.Second
+	defaultOffsetLogWriteTimeout = 3 * time.Second
+	defaultOffsetLogMaxBytes     = 8 * 1024 * 1024
+)
+
+// readOffsets satisfies the offsetStore interface, it returns a list of offsets
+// in the partition read by the log.
+func (log *offsetLog) readOffsets() ([]offset, error) {
+	c, err := log.dialLeader()
```

Comment:
We don't, we could pool the connections.

For the read operation it doesn't matter much because it only happens once when the consumer starts reading from a partition in order to sync the requeued offsets.

For the writeOffsets and deleteOffsets operations we'll probably want to reuse connections, I just wanted to get something simple working first.